dot/core/service_test.go

```diff
@@ -577,12 +657,708 @@ func Test_Service_handleBlocksAsync(t *testing.T) {
+		ctrl := gomock.NewController(t)
+		mockBlockState := NewMockBlockState(ctrl)
+		mockBlockState.EXPECT().HighestCommonAncestor(testPrevHash, testCurrentHash).
+			Return(testAncestorHash, nil)
+		mockBlockState.EXPECT().SubChain(testAncestorHash, testPrevHash).Return(testSubChain, nil)
+		mockBlockState.EXPECT().GetRuntime(nil).Return(nil, errDummyErr)
+
+		service := &Service{
+			blockState: mockBlockState,
+		}
+		execTest(t, service, testPrevHash, testCurrentHash, errDummyErr)
+	})
+
+	t.Run("invalid transaction", func(t *testing.T) {
+		t.Parallel()
+		ctrl := gomock.NewController(t)
+		runtimeMockErr := new(mocksruntime.Instance)
+		mockBlockState := NewMockBlockState(ctrl)
+		mockBlockState.EXPECT().HighestCommonAncestor(testPrevHash, testCurrentHash).
+			Return(testAncestorHash, nil)
+		mockBlockState.EXPECT().SubChain(testAncestorHash, testPrevHash).Return(testSubChain, nil)
+		mockBlockState.EXPECT().GetRuntime(nil).Return(runtimeMockErr, nil)
+		mockBlockState.EXPECT().GetBlockBody(testCurrentHash).Return(nil, errDummyErr)
+		mockBlockState.EXPECT().GetBlockBody(testAncestorHash).Return(body, nil)
+		runtimeMockErr.On("ValidateTransaction", externExt).Return(nil, errTestDummyError)
+
+		service := &Service{
+			blockState: mockBlockState,
+		}
+		execTest(t, service, testPrevHash, testCurrentHash, nil)
+	})
+
+	t.Run("happy path", func(t *testing.T) {
+		t.Parallel()
+		ctrl := gomock.NewController(t)
+		runtimeMockOk := new(mocksruntime.Instance)
+		mockBlockState := NewMockBlockState(ctrl)
+		mockBlockState.EXPECT().HighestCommonAncestor(testPrevHash, testCurrentHash).
+			Return(testAncestorHash, nil)
+		mockBlockState.EXPECT().SubChain(testAncestorHash, testPrevHash).Return(testSubChain, nil)
+		mockBlockState.EXPECT().GetRuntime(nil).Return(runtimeMockOk, nil)
+		mockBlockState.EXPECT().GetBlockBody(testCurrentHash).Return(nil, errDummyErr)
+		mockBlockState.EXPECT().GetBlockBody(testAncestorHash).Return(body, nil)
+		runtimeMockOk.On("ValidateTransaction", externExt).
+			Return(testValidity, nil)
+		mockTxnStateOk := NewMockTransactionState(ctrl)
+		mockTxnStateOk.EXPECT().AddToPool(vtx).Return(common.Hash{})
+
+		service := &Service{
+			blockState:       mockBlockState,
+			transactionState: mockTxnStateOk,
+		}
+		execTest(t, service, testPrevHash, testCurrentHash, nil)
+	})
+}
+
+func TestServiceInsertKey(t *testing.T) {
+	t.Parallel()
+	keyStore := keystore.GlobalKeystore{
+		Babe: keystore.NewBasicKeystore(keystore.BabeName, crypto.Sr25519Type),
+	}
+
+	keyring, _ := keystore.NewSr25519Keyring()
+	aliceKeypair := keyring.Alice().(*sr25519.Keypair)
+	type args struct {
+		kp           crypto.Keypair
+		keystoreType string
+	}
+	tests := []struct {
+		name      string
+		service   *Service
+		args      args
+		expErr    error
+		expErrMsg string
+	}{
+		{
+			name: "ok case",
+			service: &Service{
+				keys: &keyStore,
+			},
+			args: args{
+				kp:           aliceKeypair,
+				keystoreType: (string)(keystore.BabeName),
```

Comment:
nit You could also do the following which is more common for some reason

```suggestion
				keystoreType: string(keystore.BabeName),
```